### PR TITLE
Fix coredump deadlock with overly long backtraces

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

File: src/shared/elf-util.c [hunk 1]
```diff
@@ -860,7 +885,7 @@ int parse_elf_object(int fd, const char *executable, bool fork_disable_dump, cha
 
                 r = json_parse_file(json_in, NULL, 0, &package_metadata, NULL, NULL);
                 if (r < 0 && r != -EINVAL) /* EINVAL: json was empty, so we got nothing, but that's ok */
-                        return r;
+                        log_warning_errno(r, "Failed to read or parse json metadata, ignoring: %m");
```
Comment (yuwata) [inline src/shared/elf-util.c:888]: If the passed json data is truncated, `json_parse_file()` returns `-EINVAL`. Should we drop the condition `r != -EINVAL`?
Or, `if (r < 0 && package_metadata)` ??

**Review by yuwata — APPROVED** [review]

**Comment by AdamWill** [comment]
so I tried a systemd with this patch on a gnome-calendar crash, and still didn't get a coredump, probably because:

    Oct 18 18:24:21 fedora systemd[1]: Started systemd-coredump@0-2842-0.service - Process Core Dump (PID 2842/UID 0).
    ....
    Oct 18 18:29:21 fedora systemd[1]: systemd-coredump@0-2842-0.service: Service reached runtime time limit. Stopping.
    Oct 18 18:29:21 fedora systemd[1]: systemd-coredump@0-2842-0.service: Failed with result 'timeout'.

so it took over five minutes to deal with the core dump, and timed out...

**Comment by keszybz** [comment]
> so I tried a systemd with this patch on a gnome-calendar crash, and still didn't get a coredump, probably because:
> Oct 18 18:29:21 fedora systemd[1]: systemd-coredump@0-2842-0.service: Failed with result 'timeout'.

On my machine, generating the crash takes 5 seconds. So if it timed out, either something is wrong with the patch, or it wasn't applied properly, or there's some other bug. I'll test my patch again here.

File: src/shared/elf-util.c (hunk 1 above)
Comment (keszybz) [reply src/shared/elf-util.c:888]: That's a good point. I added another commit to change json_parse_file() to return a different error in that case.

**Comment by poettering** [comment]
Hmm, usually when reading output from a child we wait for EOF on that before waiting for the exit code of the child. Here it is done the other way round, which is the bug, no?

**Comment by keszybz** [comment]
> Hmm, usually when reading output from a child we wait for EOF on that before waiting for the exit code of the child. Here it is done the other way round, which is the bug, no?

I considered that too. We could read and accumulate input from the child until EOF. But that'd make the receiving side more complicated, because it'd need to grow a buffer, do a loop, and so on. But realistically, I don't think this is useful. The child generates the whole string first, so it knows how big it is, and can size the pipe appropriately. And also, the output should not be arbitrarily long. After all, we're going to stuff this in the journal, and we don't want to put a million lines in the message. So a more generic transfer mechanism would be more complicated but it wouldn't really give us any gains.

**Review by bluca — APPROVED** [review]

**Comment by mrc0mmand** [comment]
The C8S fail is unrelated (#25060).

**Comment by poettering** [comment]
hmm, one question: maybe the transfer mechanism should not actually be a pipe, but a memfd? a memfd can auto-grow and we wouldn't be subject to pipe size limits? i.e. instead of a pipe, just allocate a memfd, and pass it to the child? child writes its stuff there. afterwards we seek back to the front and read it?

I think the code would be more robust and even simpler?

**Comment by poettering** [comment]
memfd are 3.17 stuff, which is older than our recommended baseline, hence I think that would be Ok to simply use?

**Comment by poettering** [comment]
memfd are just memory that is accounted like other memory, so it's generally a good idea to use, since less special than memfd

**Comment by bluca** [comment]
> hmm, one question: maybe the transfer mechanism should not actually be a pipe, but a memfd? a memfd can auto-grow and we wouldn't be subject to pipe size limits? i.e. instead of a pipe, just allocate a memfd, and pass it to the child? child writes its stuff there. afterwards we seek back to the front and read it?
> 
> I think the code would be more robust and even simpler?

Probably best to leave that for the next release, and just fix the bug as-is for 252? It's a different model, so there is inherent risk of introducing new regressions by changing over

**Comment by poettering** [comment]
dunno i think switching this to memfds will remove the entire category of bug here, instead of just trying to massage it to not happen anymore.

**Comment by bluca** [comment]
> dunno i think switching this to memfds will remove the entire category of bug here, instead of just trying to massage it to not happen anymore.

Yes, but it could introduce new bugs that we aren't aware of - post-rc2 is late, surely it can wait a couple of weeks till after the next release?

**Comment by keszybz** [comment]
Centos CI (CentOS Stream 8) is TEST-15-DROPIN, i.e. #25060. Not related.

> Yes, but it could introduce new bugs that we aren't aware of - post-rc2 is late, surely it can wait a couple of weeks till after the next release?

Yep. I think we should merge this. Memfd might be nicer, but it'll likely touch a lot of code and has higher potential for regressions.

**Comment by evverx** [comment]
61aea456c12c5 got CVE-2022-45873

**Pull request merged**